src/coreclr/jit/compiler.hpp

```diff
@@ -5043,6 +5043,141 @@ void Compiler::fgVisitBlocksInLoopAwareRPO(FlowGraphDfsTree* dfsTree, FlowGraphN
     }
 }
 
+//------------------------------------------------------------------------------
+// fgRunReverseDfs: Run DFS over the reverse flow graph.
+//
+// Type parameters:
+//   VisitPreorder  - Functor type that takes a BasicBlock* and its preorder number
+//   VisitPostorder - Functor type that takes a BasicBlock* and its postorder number
+//
+// Parameters:
+//   visitPreorder  - Functor to visit block in its preorder
+//   visitPostorder - Functor to visit block in its postorder
+//   pseudoExit     - non-graph basic block to serve as the postdominator of all blocks
+//
+// Returns:
+//   Number of blocks visited.
+//
+// Notes:
+//  Requires DFS tree to be prebuilt.
+//
+template <typename VisitPreorder, typename VisitPostorder>
+unsigned Compiler::fgRunReverseDfs(VisitPreorder visitPreorder, VisitPostorder visitPostorder, BasicBlock* pseudoExit)
+{
+    BitVecTraits traits(fgBBNumMax + 1, this);
+    BitVec       visited(BitVecOps::MakeEmpty(&traits));
+
+    unsigned preOrderIndex  = 0;
+    unsigned postOrderIndex = 0;
+
+    struct PredInfo
+    {
+        PredInfo(BasicBlock* block)
+            : m_block(block)
+            , m_edge(block->bbPreds)
+        {
+        }
+        BasicBlock* m_block;
+        FlowEdge*   m_edge;
+    };
+
+    ArrayStack<PredInfo> blocks(getAllocator(CMK_DepthFirstSearch));
+
+    auto reverseDfsFrom = [&](BasicBlock* firstBB) {
+        BitVecOps::AddElemD(&traits, visited, firstBB->bbNum);
+        FlowEdge* const preds = firstBB->bbPreds;
```

Comment:
```suggestion
        FlowEdge* const preds = BlockPredsWithEH(firstBB);
```

I don't see a good reason to not model things faithfully here.